crates/sui-types/src/programmable_transaction_builder.rs

```diff
@@ -0,0 +1,244 @@
+            anyhow::bail!("blocked by gas smashing")
+        }
+        TransactionKind::Single(t) => builder.single_transaction(t)?,
+        TransactionKind::Batch(ts) => {
+            for t in ts {
+                builder.single_transaction(t)?
+            }
+        }
+    };
+    let pt = builder.finish();
+    m.kind = TransactionKind::Single(SingleTransactionKind::ProgrammableTransaction(pt));
+    Ok(m)
+}
+
+#[derive(Default)]
+pub struct ProgrammableTransactionBuilder {
+    inputs: IndexSet<CallArg>,
+    commands: Vec<Command>,
+}
+
+impl ProgrammableTransactionBuilder {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn finish(self) -> ProgrammableTransaction {
+        let Self { inputs, commands } = self;
+        let inputs = inputs.into_iter().collect();
+        ProgrammableTransaction { inputs, commands }
+    }
+
+    pub fn pure<T: Serialize>(&mut self, value: T) -> anyhow::Result<Argument> {
+        Ok(self.input(CallArg::Pure(bcs::to_bytes(&value).map_err(|e| {
+            anyhow::anyhow!("Unable to serialize pure argument. Got error: {e}")
+        })?)))
+    }
+
+    pub fn obj(&mut self, obj_arg: ObjectArg) -> Argument {
+        self.input(CallArg::Object(obj_arg))
+    }
+
+    pub fn input(&mut self, call_arg: CallArg) -> Argument {
+        assert!(!matches!(call_arg, CallArg::ObjVec(_)));
+        Argument::Input(self.inputs.insert_full(call_arg).0 as u16)
+    }
+
+    pub fn command(&mut self, command: Command) -> Argument {
+        let i = self.commands.len();
+        self.commands.push(command);
+        Argument::Result(i as u16)
+    }
+
+    pub fn single_transaction(&mut self, t: SingleTransactionKind) -> anyhow::Result<()> {
+        match t {
+            SingleTransactionKind::ProgrammableTransaction(_) => anyhow::bail!(
+                "ProgrammableTransaction are not supported in ProgrammableTransactionBuilder"
+            ),
+            SingleTransactionKind::TransferObject(TransferObject {
+                recipient,
+                object_ref,
+            }) => self.transfer_object(recipient, object_ref),
+            SingleTransactionKind::Publish(MoveModulePublish { modules }) => self.publish(modules),
+            SingleTransactionKind::Call(MoveCall {
+                package,
+                module,
+                function,
+                type_arguments,
+                arguments,
+            }) => self.move_call(package, module, function, type_arguments, arguments)?,
+            SingleTransactionKind::TransferSui(TransferSui { recipient, amount }) => {
+                self.transfer_sui(recipient, amount)
+            }
+            SingleTransactionKind::Pay(Pay {
+                coins,
+                recipients,
+                amounts,
+            }) => self.pay(coins, recipients, amounts)?,
+            SingleTransactionKind::PaySui(_) | SingleTransactionKind::PayAllSui(_) => {
+                anyhow::bail!(
```

Comment:
Is this an existing limitation of `PaySui` and `PayAllSui` (that they can't exist in a batch) or a new limitation from programmable transactions?

Reply:
You cannot map PaySui into a SingleTransacitonKind, but only TransactionData because the coins need to be in the gas vector with gas smashing